lessons/advanced/strings-in-depth.md

```diff
@@ -0,0 +1,163 @@
+---
+layout: page
+title: Strings in-depth 
+category: advanced
+order: 9
+lang: en
+---
+
+String manipulation is one of our must to have skills nowadays, ignoring the programming language. That's why we are including an in-depth on Strings.
+
+## Table of Contents
+
+- [Strings in Elixir](#strings-in-elixir)
+- [Char lists](#char-lists)
+- [Graphemes and Codepoints](#graphemes)
+- [String functions](#string-functions)
+  - [Length](#length)
+  - [Replace](#replace)
+  - [Duplicate](#duplicate)
+- [Exercises](#exercises)
+  - [Anagrams](#anagrams)
+
+## Strings in Elixir
+
+Elixir strings are nothing but a sequence of bytes. Let's watch a clear example:
+
+```elixir
+iex(2)> string = <<104,101,108,108,111>>
+"hello"
+```
+
+>NOTE: Using << >> syntax we are saying to the compiler that the elements inside those symbols are bytes.
+
+On the previous example we just defined a string with a list of bytes, and the clear result was the string "hello". If we want to use a Guard in our functions to check if a parameter is a string, we just have to write:
+
+```elixir
+def parse_string(string) when is_binary(string) do
+  ...
+end
+```
+
+## Char lists
+
+Internally, Elixir strings are represented with a sequence of bytes rather than an array of characters, but it's worth to mention Elixir also has a data type that is a char list (characters list).
+
+>NOTE: Elixir strings are created with double quotes, while char lists are with single quotes.
+
+What's the difference between them? Each value from a char list is the ASCII value from the character. Let's dig in:
+
+```elixir
+iex(12)> string = 'hello'
+'hello'
+iex(14)> Enum.reduce(string, "", fn char, acc -> acc <> to_string(char) <> "," end)
+"104,101,108,108,111,"
+```
+
+While programming in Elixir, we are not usually using char lists but Strings. The char lists support is given because are required by some Erlang modules.
+
+## Graphemes and codepoints
+
+Codepoints are just simple Unicode characters, which may be represented by one or two bytes. For example, characters with a tilde or accents: `á, ñ, è`. Graphemes consists on multiple codepoints that look as a simple character.
+
+The String module already provides two methods to obtain them, `graphemes/1` and `codepoints/1`. Let's look at the example:
+
+```elixir
+iex(14)> string = "\u0061\u0301"
+"á"
+iex(15)> String.codepoints string
+["a", "́"]
+iex(16)> String.graphemes string
+["á"]
+```
+
+## String functions
+
+Let's review some of the most important and useful function the String module has for us.
+
+### `length/1`
+
+Returns the number of Graphemes in the string.
+
+```elixir
+iex> String.length "Hello"
+5
+```
+
+### `replace(subject, pattern, replacement, options \\ [])`
```

Comment:
We should be consistent throughout so let's use the function arity for all of these rather than switching between `length/1` and `replace(subject, pattern, replacement, options \\ [])` formatting.